serializer/stream/read.go

```diff
@@ -2,102 +2,158 @@ package stream
+	if err != nil {
+		return nil, ierrors.Wrap(err, "failed to read serialized bytes")
+	}
+	if nBytes != length {
+		return nil, ierrors.Errorf("failed to read serialized bytes: read bytes (%d) != size (%d)", nBytes, length)
 	}
 
-	if consumedBytes, err := target.FromBytes(readBytes); err != nil {
-		return ierrors.Wrap(err, "failed to parse bytes of serializable")
-	} else if len(optFixedSize) > 1 && consumedBytes != len(readBytes) {
-		return ierrors.Errorf("failed to parse serializable: consumed bytes (%d) != read bytes (%d)", consumedBytes, len(readBytes))
+	return readBytes, nil
+}
+
+// ReadBytesWithSize reads a byte slice from the reader where lenType specifies the serialization length prefix type.
+func ReadBytesWithSize(reader io.Reader, lenType serializer.SeriLengthPrefixType) ([]byte, error) {
+	size, err := readFixedSize(reader, lenType)
+	if err != nil {
+		return nil, ierrors.Wrap(err, "failed to read blob size")
 	}
 
-	return
+	return ReadBytes(reader, size)
 }
 
-// ReadFunc reads a type from the stream as specified by readFunc. If the serialized type is of fixed size, we can provide
-// the length to omit additional information about the length.
-// Note: optFixedSize needs to be specified when writing this type as well, otherwise there will be issues with the bytes' alignment.
-func ReadFunc[T any](reader io.ReadSeeker, readFunc func(bytes []byte) (T, int, error), optFixedSize ...int) (T, error) {
-	var readBytes []byte
+// ReadObject reads a type from the reader as specified by objectFromBytesFunc. A fixed length for the deserialized type must be specified.
+func ReadObject[T any](reader io.Reader, fixedLen int, objectFromBytesFunc func(bytes []byte) (T, int, error)) (T, error) {
 	var result T
-	var err error
 
-	if len(optFixedSize) == 0 {
-		if readBytes, err = ReadBlob(reader); err != nil {
-			return result, ierrors.Wrap(err, "failed to read serialized bytes")
-		}
-	} else {
-		if readBytes, err = ReadBytes(reader, uint64(optFixedSize[0])); err != nil {
-			return result, ierrors.Wrap(err, "failed to read serialized bytes")
-		}
+	readBytes, err := ReadBytes(reader, fixedLen)
+	if err != nil {
+		return result, ierrors.Wrap(err, "failed to read serialized bytes")
+	}
+
+	result, consumedBytes, err := objectFromBytesFunc(readBytes)
+	if err != nil {
+		return result, ierrors.Wrap(err, "failed to parse bytes of objectFromBytesFunc")
+	}
+	if consumedBytes != len(readBytes) {
+		return result, ierrors.Errorf("failed to parse objectFromBytesFunc: consumed bytes (%d) != read bytes (%d)", consumedBytes, len(readBytes))
+	}
+
+	return result, nil
+}
+
+// ReadObjectWithSize reads a type from the reader as specified by fromBytesFunc. The serialization length prefix type must be specified.
+func ReadObjectWithSize[T any](reader io.Reader, lenType serializer.SeriLengthPrefixType, objectFromBytesFunc func(bytes []byte) (T, int, error)) (T, error) {
+	var result T
+
+	readBytes, err := ReadBytesWithSize(reader, lenType)
```

Comment:
why not call `readFixedSize` and then `ReadObject` ? It would reduce the duplicate code